File: Work_leetcode/Contests/week 13/matchstick.py
class Solution(object):
    def makesquare(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        sum = [0,0,0,0]
        s = 0
        for i in nums:
            s += i
        if s % 4 != 0 or s == 0:
            return False
        subsum = s / 4
        nums.sort(reverse = True)
        return self.helper(0, nums, len(nums), subsum, sum)
    def helper(self, start, nums, len, subsum, sum):
        if start == len:
            for i in sum:
                if i != subsum:
                    return False
            return True
        else:
            subres = False
            for i in xrange(4):
                if sum[i] + nums[start] > subsum: continue
                sum[i] += nums[start]
                if (self.helper(start+1, nums, len, subsum, sum)):
                    subres = True
                    break
                sum[i] -= nums[start]
            return subres
        return True
        
        

File: Work_leetcode/Contests/week 13/matchstick.java
public class Solution {
    public boolean makesquare(int[] nums) {
        int[] sum = new int[4];
        sum[0] = sum[1] = sum[2] = sum[3] = 0;
        int s = 0;
        for (int num : nums) s += num;
        if (s % 4 != 0 || s == 0) return false;
        int subsum = s / 4;
        Arrays.sort(nums);
        return helper(nums.length - 1, nums, subsum, sum);
    }
    public boolean helper(int start, int[] nums, int subsum, int[] sum) {
        if (start == -1) {
            for (int i : sum) {
                if (i != subsum) return false;
            }
            return true;
        } else {
            boolean subres = false;
            for (int i=0; i<4; ++i) {
                if (sum[i] + nums[start] > subsum) continue;
                sum[i] += nums[start];
                if (helper(start-1, nums, subsum, sum)) {subres = true; break;}
                sum[i] -= nums[start];
            }
            return subres;
        }
    }
}